src/connection.rs

```diff
@@ -8,793 +8,841 @@
-            return Err(ConnectionError::TooManyPendingFrames)
+    /// Get the next inbound `Stream` and make progress along the way.
+    ///
+    /// This is called from `Connection::next_stream` instead of being a
+    /// public method itself in order to guarantee proper closing in
+    /// case of an error or at EOF.
+    async fn next(&mut self) -> Result<Option<Stream>> {
+        loop {
+            self.garbage_collect().await?;
+
+            // For each channel and the socket we create a future that gets
+            // the next item. We will poll each future and if any one of them
+            // yields an item, we return the tuple of poll results which are
+            // then all processed.
+            //
+            // For terminated sources we create non-finishing futures.
+            // This guarantees that if the remaining futures are pending
+            // we properly wait until woken up because we actually can make
+            // progress.
+            //
+            // While it should never happen that all futures are terminated
+            // we nevertheless count them and return early if they are.
+
+            let mut num_terminated = 0;
+
+            let mut next_inbound_frame =
+                if self.socket.is_terminated() {
+                    num_terminated += 1;
+                    Either::Left(future::pending())
+                } else {
+                    Either::Right(self.socket.try_next().err_into())
+                };
+
+            let mut next_stream_command =
+                if self.stream_receiver.is_terminated() {
+                    num_terminated += 1;
+                    Either::Left(future::pending())
+                } else {
+                    Either::Right(self.stream_receiver.next())
+                };
+
+            let mut next_control_command =
+                if self.control_receiver.is_terminated() {
+                    num_terminated += 1;
+                    Either::Left(future::pending())
+                } else {
+                    Either::Right(self.control_receiver.next())
+                };
+
+            if num_terminated == 3 {
+                log::error!("{}: all futures are terminated", self.id);
+                return Err(ConnectionError::Closed)
+            }
+
+            let next_item =
+                future::poll_fn(move |cx: &mut Context| {
+                    let a = next_stream_command.poll_unpin(cx);
+                    let b = next_control_command.poll_unpin(cx);
+                    let c = next_inbound_frame.poll_unpin(cx);
+                    if a.is_pending() && b.is_pending() && c.is_pending() {
+                        return Poll::Pending
+                    }
+                    Poll::Ready((a, b, c))
+                });
+
+            let (stream_command, control_command, inbound_frame) = next_item.await;
+
+            if let Poll::Ready(cmd) = control_command {
+                self.on_control_command(cmd).await?
```

Comment:
This is a bit dangerous. If the future is interrupted (i.e. the user polls the future returned by `next()` then destroys that future before it ends), then we are losing the `stream_command` and `inbound_frame` (and might end up with an inconsistent state?).

I might be a better idea to store the pending commands in the `Connection` across yield points, rather than local variables.

Reply:
As discussed in person, we will not support cancellation of `Connection::next_stream` for the time being. A  comment [was added](https://github.com/paritytech/yamux/pull/59/commits/3dd2f969073f0ecd649f09a68e507a51e8cbe895) to make this explicit.

Reply:
While it would not be surprising to me that things may break if futures returned from `next_stream` are not polled to completion given that `next_stream` is the overall driving force of the connection, it would just be good to clearly document that fact.